[clang-tidy] Warn about misuse of sizeof operator in loops. (#143205)
llvm/llvm-project: clang-tools-extra/docs/ReleaseNotes.rst

  <clang-tidy/checks/bugprone/optional-value-conversion>` check to detect
  conversion in argument of ``std::make_optional``.

- Improved :doc: `bugprone-sizeof-expression
  <clang-tidy/checks/bugprone/bugprone-sizeof-expression>` check by adding
  `WarnOnSizeOfInLoopTermination` option to detect misuses of ``sizeof``

>   expression in loop conditions. The check detects the use of ``sizeof``

Reviewer: One statement is enough.

  tolerating fix-it breaking compilation when functions is used as pointers
  to avoid matching usage of functions within the current compilation unit.

> - Improved :doc: `bugprone-sizeof-expression

Author: Done.

>  <clang-tidy/checks/bugprone/bugprone-sizeof-expression>` check.
>  Introduced WarnOnSizeOfInLoopTermination sub-class to detect misuses of sizeof

Reviewer: ```suggestion
 <clang-tidy/checks/bugprone/bugprone-sizeof-expression>` check by adding
 `WarnOnSizeOfInLoopTermination` option to detect misuses of ``sizeof``
```

  tolerating fix-it breaking compilation when functions is used as pointers
  to avoid matching usage of functions within the current compilation unit.

- Improved :doc: `bugprone-sizeof-expression
 <clang-tidy/checks/bugprone/bugprone-sizeof-expression>` check.

>  Introduced WarnOnSizeOfInLoopTermination sub-class to detect misuses of sizeof expression

Reviewer: ```suggestion
 Introduced `WarnOnSizeOfInLoopTermination` option to detect misuses of ``sizeof`` expression
```

>  Introduced WarnOnSizeOfInLoopTermination sub-class to detect misuses of sizeof expression

Reviewer 2: Please make lines no more than 80 characters long
